### PR TITLE
Fix a bug where nodes that weren't schedulable or ready are added to load balancers

### PR DESCRIPTION
**Pull request by brendandburns**
Ref #16438, #13346, #16278 

@fgrzadkowski @davidopp

### CODE REVIEW

File: pkg/controller/service/servicecontroller.go [hunk 1]
```diff
@@ -581,9 +581,15 @@ func stringSlicesEqual(x, y []string) bool {
...
 
 func hostsFromNodeList(list *api.NodeList) []string {
-	result := make([]string, len(list.Items))
+	result := []string{}
 	for ix := range list.Items {
-		result[ix] = list.Items[ix].Name
+		if list.Items[ix].Spec.Unschedulable {
+			continue
+		}
+		if list.Items[ix].Status.Phase != api.NodeRunning {
```
Comment (bprashanth) [inline pkg/controller/service/servicecontroller.go]: heh, I ran into the same problem with L7. you want not ready also no, just use the condition lister https://github.com/bprashanth/contrib/commit/4dc30b3de54dbdf9be9a978589e854d577231ffa#diff-315afc43468bd4edd63930333434d306R358 ?

**Comment by davidopp** [comment]
LGTM

But you assigned it to Clayton, so I'll let him add the label.

Thanks for the fix!

**Comment by davidopp** [comment]
@k8s-bot test this please

File: pkg/controller/service/servicecontroller.go (hunk 1 above)
Comment (brendandburns) [inline pkg/controller/service/servicecontroller.go]: done.

**Comment by brendandburns** [comment]
Updated to reflect @bprashanth 's comments.

@davidopp the merge-bot is the one that actually assigned things, feel free to mark LGTM if you want.

Thanks!
--brendan

**Comment by davidopp** [comment]
Oh sorry, I thought you had assigned it to Clayton, didn't notice it was mergebot.

Labeling now.

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Comment by wojtek-t** [comment]
Jenkins failure:
"_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/service/servicecontroller.go:604: too many arguments in call to s.nodeLister.NodeCondition"

@brendandburns - can you please rebase?

**Comment by brendandburns** [comment]
Updated and rebased.

**Comment by davidopp** [comment]
LGTM

**Comment by davidopp** [comment]
Hey, I just want to make sure you understand that
s.nodeLister.NodeCondition(getNodeConditionPredicate()).List()
will only give you back the nodes that the scheduler considers schedulable. Is that what you want? (You changed s.NodeLister.List() to s.nodeLister.NodeCondition(getNodeConditionPredicate()).List()
in one spot.)

Temporarily removed LGTM

**Comment by davidopp** [comment]
Sorry, my comment was braindead, I forgot that you are defining your own predicate here, not reusing the one from the scheduler.

**Comment by brendandburns** [comment]
Added a Unschedulable to the predicate and added a test.

ptal.

thanks!
--brendan

**Comment by davidopp** [comment]
LGTM

**Comment by wojtek-t** [comment]
@k8s-bot unit test this please

**Pull request merged**